src/lib/core/GroupedCallbackList.h

```diff
@@ -0,0 +1,227 @@
+    // both (a, nullptr) and (nullptr, b) are by necessity represented by only
+    // a single node in the list. To be able to recover this information, we use
+    // distinct trampolines that tag the "cancel" function pointer stored in each
+    // node with the index of the callback type within the argument type tuple.
+
+    template <std::size_t... Index>
+    void Enqueue(std::index_sequence<Index...>, Callback<T> *... callback)
+    {
+        Cancelable * group = nullptr;
+        (
+            [&] {
+                VerifyOrReturn(callback != nullptr);
+                detail::EnqueueWithGroup(callback->Cancel(), group, this, &detail::TaggedDequeueGroup<Index>);
+            }(),
+            ...);
+    }
+
+    template <std::size_t... Index>
+    bool Peek(std::index_sequence<Index...>, Callback<T> *&... callback) const
+    {
+        Cancelable * cancelable = mNext;
+        VerifyOrReturnValue(cancelable != this, false);
+        Cancelable * groupPrev = cancelable->mPrev;
+        (
+            [&] {
+                if (cancelable->mPrev == groupPrev && cancelable->mCancel == &detail::TaggedDequeueGroup<Index>)
+                {
+                    callback   = Callback<decltype(callback->mCall)>::FromCancelable(cancelable);
+                    cancelable = cancelable->mNext;
+                }
+                else
+                {
+                    callback = nullptr;
+                }
+            }(),
+            ...);
+        return true;
+    }
+};
+
+namespace detail {
+
+inline void EnqueueWithGroup(Cancelable * cancelable, Cancelable *& group, Cancelable * anchor, void (*cancelFn)(Cancelable *))
+{
+    cancelable->mCancel = cancelFn;
+    cancelable->mNext   = anchor;
+    if (!group)
+    {
+        group             = cancelable;
+        cancelable->mPrev = anchor->mPrev;
+    }
+    else
+    {
+        cancelable->mPrev = group->mPrev;
+    }
+    anchor->mPrev->mNext = cancelable;
+    anchor->mPrev        = cancelable;
+}
+
+inline void LinkGroup(Cancelable * prev, Cancelable * cancelable)
+{
+    prev->mNext = cancelable;
+
+    Cancelable * groupPrev = cancelable->mPrev;
+    do
+    {
+        cancelable->mPrev = prev;
+        cancelable        = cancelable->mNext;
+    } while (cancelable->mPrev == groupPrev);
+}
+
+inline Cancelable * ClearCancelable(Cancelable * cancelable)
+{
+    auto * next       = cancelable->mNext;
+    cancelable->mPrev = cancelable->mNext = cancelable;
+    cancelable->mCancel                   = nullptr;
+    return next;
+}
+
+inline void DequeueGroup(Cancelable * cancelable)
+{
+    Cancelable * prev = cancelable->mPrev;
+    Cancelable * next = prev->mNext;
+    do
+    {
+        next = ClearCancelable(next);
+    } while (next->mPrev == prev);
+    LinkGroup(prev, next);
+}
+
+template <size_t Index>
+void TaggedDequeueGroup(Cancelable * cancelable)
+{
+    (void) Index; // not used, we only care that instantiations have unique addresses
```

Comment:
Hmm.  Will they have unique addresses even in the face of identical code folding or whatnot?   I guess since we are taking the addresses of the functions the as-if rule should prevent them from being folded?

Reply:
I believe taking account of whether addresses get taken is considered "safe" ICF, so I think the expectation is that compilers do this correctly. We do have the unit tests that will fail if it doesn't work.